### PR TITLE
r/aws_spot_fleet_request - add support for tagging + validations

### LINKED ISSUE #12078: r/aws_spot_fleet_request: Support tagging
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

EC2 Spot Fleet request can now be tagged.
These tags apply to the spot fleet request resource and are separate from the tags that can be specified in the `launch_specification` configuration block and which apply to the EC2 instances in the fleet. 

### New or Affected Resource(s)

* [`aws_spot_fleet_request`](https://www.terraform.io/docs/providers/aws/r/spot_fleet_request.html)

### Potential Terraform Configuration

```hcl
resource "aws_spot_fleet_request" "example" {
  tags = {
    Usage = "dev"
  }
}
```

### References

[Announcement](https://aws.amazon.com/about-aws/whats-new/2020/02/amazon-ec2-now-supports-tagging-ec2-spot-fleet-requests/).
[API reference](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SpotFleetRequestConfigData.html).

Requires [AWS SDK v1.29.0](https://github.com/aws/aws-sdk-go/releases/tag/v1.29.0):
* https://github.com/terraform-providers/terraform-provider-aws/pull/11893

Related:
* https://github.com/terraform-providers/terraform-provider-aws/pull/11934

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #12078

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_spot_fleet_request: add support for tagging.
resource_aws_spot_fleet_request: add plan time validation to `volume_type`, `iam_instance_profile_arn`, `allocation_strategy`, `instance_interruption_behaviour`, `target_group_arns`, `placement_tenancy`.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSpotFleetRequest_'
--- PASS: TestAccAWSSpotFleetRequest_basic (344.44s)
--- PASS: TestAccAWSSpotFleetRequest_updateTargetCapacity (986.78s)
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion (364.98s)
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList (349.96s)
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList (295.24s)
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz (291.23s)
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet (302.79s)
--- PASS: TestAccAWSSpotFleetRequest_overriddingSpotPrice (281.26s)
--- PASS: TestAccAWSSpotFleetRequest_withoutSpotPrice (341.80s)
--- PASS: TestAccAWSSpotFleetRequest_diversifiedAllocation (372.79s)
--- PASS: TestAccAWSSpotFleetRequest_multipleInstancePools (362.38s)
--- PASS: TestAccAWSSpotFleetRequest_withWeightedCapacity (358.54s)
--- PASS: TestAccAWSSpotFleetRequest_withEBSDisk (313.12s)
--- PASS: TestAccAWSSpotFleetRequest_LaunchSpecification_EbsBlockDevice_KmsKeyId (213.91s)
--- PASS: TestAccAWSSpotFleetRequest_LaunchSpecification_RootBlockDevice_KmsKeyId (227.32s)
--- PASS: TestAccAWSSpotFleetRequest_withTags (354.02s)
--- PASS: TestAccAWSSpotFleetRequest_placementTenancyAndGroup (106.36s)
--- PASS: TestAccAWSSpotFleetRequest_WithELBs (364.46s)
--- PASS: TestAccAWSSpotFleetRequest_WithTargetGroups (497.25s)
--- PASS: TestAccAWSSpotFleetRequest_associatePublicIpAddress (340.35s)
--- PASS: TestAccAWSSpotFleetRequest_instanceInterruptionBehavior (277.50s)
--- PASS: TestAccAWSSpotFleetRequest_fleetType (272.67s)
--- PASS: TestAccAWSSpotFleetRequest_iamInstanceProfileArn (332.60s)
--- PASS: TestAccAWSSpotFleetRequest_changePriceForcesNewRequest (480.32s)
--- PASS: TestAccAWSSpotFleetRequest_tags (346.79s)
```

### CODE REVIEW

**Comment by ewbankkit** [comment]
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSpotFleetRequest_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSSpotFleetRequest_ -timeout 120m
=== RUN   TestAccAWSSpotFleetRequest_basic
=== PAUSE TestAccAWSSpotFleetRequest_basic
=== RUN   TestAccAWSSpotFleetRequest_tags
=== PAUSE TestAccAWSSpotFleetRequest_tags
=== RUN   TestAccAWSSpotFleetRequest_associatePublicIpAddress
=== PAUSE TestAccAWSSpotFleetRequest_associatePublicIpAddress
=== RUN   TestAccAWSSpotFleetRequest_instanceInterruptionBehavior
=== PAUSE TestAccAWSSpotFleetRequest_instanceInterruptionBehavior
=== RUN   TestAccAWSSpotFleetRequest_fleetType
=== PAUSE TestAccAWSSpotFleetRequest_fleetType
=== RUN   TestAccAWSSpotFleetRequest_iamInstanceProfileArn
=== PAUSE TestAccAWSSpotFleetRequest_iamInstanceProfileArn
=== RUN   TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
=== PAUSE TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
=== RUN   TestAccAWSSpotFleetRequest_updateTargetCapacity
=== PAUSE TestAccAWSSpotFleetRequest_updateTargetCapacity
=== RUN   TestAccAWSSpotFleetRequest_updateExcessCapacityTerminationPolicy
=== PAUSE TestAccAWSSpotFleetRequest_updateExcessCapacityTerminationPolicy
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
=== PAUSE TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
=== PAUSE TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
=== RUN   TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
=== PAUSE TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
=== PAUSE TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
=== RUN   TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
=== PAUSE TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
=== RUN   TestAccAWSSpotFleetRequest_overriddingSpotPrice
=== PAUSE TestAccAWSSpotFleetRequest_overriddingSpotPrice
=== RUN   TestAccAWSSpotFleetRequest_withoutSpotPrice
=== PAUSE TestAccAWSSpotFleetRequest_withoutSpotPrice
=== RUN   TestAccAWSSpotFleetRequest_diversifiedAllocation
=== PAUSE TestAccAWSSpotFleetRequest_diversifiedAllocation
=== RUN   TestAccAWSSpotFleetRequest_multipleInstancePools
=== PAUSE TestAccAWSSpotFleetRequest_multipleInstancePools
=== RUN   TestAccAWSSpotFleetRequest_withWeightedCapacity
=== PAUSE TestAccAWSSpotFleetRequest_withWeightedCapacity
=== RUN   TestAccAWSSpotFleetRequest_withEBSDisk
=== PAUSE TestAccAWSSpotFleetRequest_withEBSDisk
=== RUN   TestAccAWSSpotFleetRequest_LaunchSpecification_EbsBlockDevice_KmsKeyId
=== PAUSE TestAccAWSSpotFleetRequest_LaunchSpecification_EbsBlockDevice_KmsKeyId
=== RUN   TestAccAWSSpotFleetRequest_LaunchSpecification_RootBlockDevice_KmsKeyId
=== PAUSE TestAccAWSSpotFleetRequest_LaunchSpecification_RootBlockDevice_KmsKeyId
=== RUN   TestAccAWSSpotFleetRequest_withTags
=== PAUSE TestAccAWSSpotFleetRequest_withTags
=== RUN   TestAccAWSSpotFleetRequest_placementTenancyAndGroup
=== PAUSE TestAccAWSSpotFleetRequest_placementTenancyAndGroup
=== RUN   TestAccAWSSpotFleetRequest_WithELBs
=== PAUSE TestAccAWSSpotFleetRequest_WithELBs
=== RUN   TestAccAWSSpotFleetRequest_WithTargetGroups
=== PAUSE TestAccAWSSpotFleetRequest_WithTargetGroups
=== RUN   TestAccAWSSpotFleetRequest_WithInstanceStoreAmi
--- SKIP: TestAccAWSSpotFleetRequest_WithInstanceStoreAmi (0.00s)
    resource_aws_spot_fleet_request_test.go:768: Test fails due to test harness constraints
=== CONT  TestAccAWSSpotFleetRequest_basic
=== CONT  TestAccAWSSpotFleetRequest_overriddingSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_basic (274.45s)
=== CONT  TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet
--- PASS: TestAccAWSSpotFleetRequest_overriddingSpotPrice (309.23s)
=== CONT  TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameSubnet (244.49s)
=== CONT  TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList
--- PASS: TestAccAWSSpotFleetRequest_multipleInstanceTypesInSameAz (276.82s)
=== CONT  TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceSubnetInGivenList (274.27s)
=== CONT  TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzInGivenList (276.85s)
=== CONT  TestAccAWSSpotFleetRequest_updateExcessCapacityTerminationPolicy
--- PASS: TestAccAWSSpotFleetRequest_lowestPriceAzOrSubnetInRegion (294.79s)
=== CONT  TestAccAWSSpotFleetRequest_updateTargetCapacity
--- PASS: TestAccAWSSpotFleetRequest_updateExcessCapacityTerminationPolicy (570.15s)
=== CONT  TestAccAWSSpotFleetRequest_changePriceForcesNewRequest
--- PASS: TestAccAWSSpotFleetRequest_changePriceForcesNewRequest (477.54s)
=== CONT  TestAccAWSSpotFleetRequest_iamInstanceProfileArn
--- PASS: TestAccAWSSpotFleetRequest_updateTargetCapacity (896.16s)
=== CONT  TestAccAWSSpotFleetRequest_fleetType
--- PASS: TestAccAWSSpotFleetRequest_iamInstanceProfileArn (284.87s)
=== CONT  TestAccAWSSpotFleetRequest_instanceInterruptionBehavior
--- PASS: TestAccAWSSpotFleetRequest_fleetType (274.35s)
=== CONT  TestAccAWSSpotFleetRequest_associatePublicIpAddress
--- PASS: TestAccAWSSpotFleetRequest_instanceInterruptionBehavior (263.58s)
=== CONT  TestAccAWSSpotFleetRequest_tags
--- PASS: TestAccAWSSpotFleetRequest_associatePublicIpAddress (273.09s)
=== CONT  TestAccAWSSpotFleetRequest_LaunchSpecification_RootBlockDevice_KmsKeyId
--- PASS: TestAccAWSSpotFleetRequest_LaunchSpecification_RootBlockDevice_KmsKeyId (167.28s)
=== CONT  TestAccAWSSpotFleetRequest_WithTargetGroups
--- PASS: TestAccAWSSpotFleetRequest_tags (278.02s)
=== CONT  TestAccAWSSpotFleetRequest_WithELBs
--- PASS: TestAccAWSSpotFleetRequest_WithELBs (288.81s)
=== CONT  TestAccAWSSpotFleetRequest_placementTenancyAndGroup
--- PASS: TestAccAWSSpotFleetRequest_placementTenancyAndGroup (70.96s)
=== CONT  TestAccAWSSpotFleetRequest_withTags
--- PASS: TestAccAWSSpotFleetRequest_WithTargetGroups (410.57s)
=== CONT  TestAccAWSSpotFleetRequest_withWeightedCapacity
--- PASS: TestAccAWSSpotFleetRequest_withTags (240.99s)
=== CONT  TestAccAWSSpotFleetRequest_LaunchSpecification_EbsBlockDevice_KmsKeyId
--- PASS: TestAccAWSSpotFleetRequest_withWeightedCapacity (315.53s)
=== CONT  TestAccAWSSpotFleetRequest_withEBSDisk
--- PASS: TestAccAWSSpotFleetRequest_LaunchSpecification_EbsBlockDevice_KmsKeyId (155.60s)
=== CONT  TestAccAWSSpotFleetRequest_diversifiedAllocation
--- PASS: TestAccAWSSpotFleetRequest_withEBSDisk (316.07s)
=== CONT  TestAccAWSSpotFleetRequest_multipleInstancePools
--- PASS: TestAccAWSSpotFleetRequest_diversifiedAllocation (349.17s)
=== CONT  TestAccAWSSpotFleetRequest_withoutSpotPrice
--- PASS: TestAccAWSSpotFleetRequest_multipleInstancePools (246.58s)
--- PASS: TestAccAWSSpotFleetRequest_withoutSpotPrice (213.32s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	4055.989s
```

File: aws/resource_aws_spot_fleet_request.go [hunk 1]
```diff
@@ -380,9 +410,13 @@ func resourceAwsSpotFleetRequest() *schema.Resource {
 				Optional: true,
 				Computed: true,
 				ForceNew: true,
-				Elem:     &schema.Schema{Type: schema.TypeString},
-				Set:      schema.HashString,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validateArn,
```
Comment (ewbankkit) [inline aws/resource_aws_spot_fleet_request.go]: I'm not sure if this validation function will ever get called.

File: aws/resource_aws_spot_fleet_request_test.go [hunk 2]
```diff
@@ -889,6 +937,10 @@ func testAccPreCheckAWSEc2SpotFleetRequest(t *testing.T) {
 
 func testAccAWSSpotFleetRequestConfigBase(rName string, rInt int) string {
 	return fmt.Sprintf(`
+data "aws_availability_zones" "available" {
+  state = "available"
+}
```
Comment (bflad) [inline aws/resource_aws_spot_fleet_request_test.go:942]: After the creation of this PR, we enabled support for Local Zone management in the provider and in the testing of doing so in our testing accounts discovered that this data source was returning `us-west-2-lax-1a` as the first result, so we've been failures with missing instance types and functionality. Going to swap this here for the configuration that excludes Local Zones before running the testing.

```suggestion
data "aws_availability_zones" "available" {
  state = "available"

  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

This extra little configuration is now also in the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#acceptance-testing-guidelines) so you can find it there in the future if you need to. 👍

**Comment by DrFaust92** [comment]
@bflad, sure thing ill take a look at #4866. i need something fun after the horrible FSx PRs (they forever to create!)

**Comment by bflad** [comment]
@DrFaust92 you rock. Thanks. Honestly feel free to do whatever on top of the original contributor's PR to get it up to modern standards and fixing things per usual. ❤️

**Comment by DrFaust92** [comment]
missed that this closes also #6627

**Pull request merged**